### PR TITLE
Respect port provided by the client && Switch from urllib.urlparse to simple substring replacement.

### LINKED ISSUE #10359: Respect explicitely provided ports by clients in PortForward
**Impact of the new feature**
Everything

**Is your feature request related to a problem? Please describe.**
In the recently developed portForward decorator and class [1] we do not respect if a client has explicitly provided a port for the url to be mangled. In such cases it would be wise to respect the clients wish, given that  they may have a solid reason for calling the specific port they provide.  

[1]
https://github.com/dmwm/WMCore/pull/10330

**Describe the solution you'd like**
While parsing the url to be mangled, check the presence of a port parameter. If it is there, avoid changing it.

**Describe alternatives you've considered**
No alternatives. This is something I consider a flaw in the original design. Thanks to others for spoting it. 
@belforte @vkuznet @amaltaro  https://github.com/dmwm/WMCore/pull/10330#issuecomment-803018864

**Additional context**

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #10359 

#### Status
ready

#### Description
While we developing the port forwarding logic, we did not take into account the cases when the client actually provides a port as part of the url. Along the way few measurements happened regarding the efficiency of the two approaches:
   * Using the `urllib.urlparse` to parse the whole url in components and just change what is needed
   * Using the simplest possible sub string substitution. 

The current change switches back to the sub string replacement  mechanism, which implicitly satisfies the major requirement for the issue - Not to change ports provided  with the url by the client.  

#### Is it backward compatible (if not, which system it affects?)
YES 

#### Related PRs
No

#### External dependencies / deployment changes
No

### CODE REVIEW

**Comment by todor-ivanov** [comment]
Hi @amaltaro,  This is a change  enclosed completely in the `PortForward` file in `Utils`. I tried to not touch the already tested code. But in case you think (based on the measurements we did and pasted in the issue) we'd better completely scratch the decorator, I am ready to do so.

**Review by mapellidario — CHANGES_REQUESTED** [review]
I do like the idea of getting rid of `urlllib.parse`, kudos!

However I think I missed that `bytes` is not imported when reviewing #10330 and added some comments. 

Thanks Todor, sorry for disturbing with an unrequested review!

File: src/python/Utils/PortForward.py [hunk 1]
```diff
@@ -10,9 +10,6 @@
 
 from builtins import str
 
-import logging
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/Utils/PortForward.py:13]: fixed

File: src/python/Utils/PortForward.py [hunk 2]
```diff
@@ -57,40 +51,21 @@ def portMangle(callObj, url, *args, **kwargs):
...
             try:
-                oldUrl = urlparse(url)
-                found = False
                 if isinstance(url, str):
-                    for mUrl in urlMangleList:
-                        if url.startswith(mUrl):
-                            netlocStr = u'%s:%d' % (oldUrl.hostname, port)
-                            found = True
-                            break
+                    urlToMangle = 'https://cmsweb'
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/Utils/PortForward.py:57]: fixed

File: src/python/Utils/PortForward.py [hunk 3]
```diff
@@ -10,9 +10,6 @@
 
 from builtins import str
 
-import logging
-from urllib.parse import urlparse, ParseResult
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/Utils/PortForward.py:14]: fixed

File: src/python/Utils/PortForward.py [hunk 4]
```diff
@@ -10,9 +10,6 @@
 
 from builtins import str
```
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/Utils/PortForward.py:8]: Now this one I do not get it. Why should I import `bytes`? IIUC This should be a builtin for both python2 and 3? It is only that in python2 it relates to `str` by default, while in python3 it is a separate type [1] && [2]. Am I missing something? 

The thing, that actually made the magic and made it run was the import of `from builtins import str`, which helped for `u''` to be recognized as a normal string in python2 and let me get rid of checks like `isinstance(u'', (str, unicode))` - something that without a dirty aliasing and hacks would never work in python3. But other than that i do not see a reason to explicitely import anything else. 


[1]
```
$ ipython
Python 2.7.5 (default, Nov 16 2020, 22:23:17) 
Type "copyright", "credits" or "license" for more information.

IPython 5.10.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: isinstance('', bytes)
Out[1]: True

In [2]: isinstance(b'', bytes)
Out[2]: True
```

[2]
```
$ ipython
Python 3.6.8 (default, Nov 16 2020, 16:55:22) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: isinstance('', bytes)
Out[1]: False

In [2]: isinstance(b'', bytes)
Out[2]: True

```

[3]
```
$ ipython
Python 2.7.5 (default, Nov 16 2020, 22:23:17) 
Type "copyright", "credits" or "license" for more information.

IPython 5.10.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: isinstance(u'', str)
Out[1]: False

In [2]: from builtins import str

In [3]: isinstance(u'', str)
Out[3]: True

```

**Review by amaltaro — COMMENTED** [review]
Todor, once you fix those unit tests, make another review request please.

**Comment by todor-ivanov** [comment]
They are fixed @amaltaro , It is only the jenkins run that  we are waiting for.

**Comment by todor-ivanov** [comment]
@amaltaro I cannot figure out why this one fails:

```
WMCore_t.DataStructs_t.LumiList_t.LumiListTest:testURL changed from success to error
```

I simply fail to see how my changes affect it.

**Comment by amaltaro** [comment]
Todor, I would suggest you to look at what gets reported in jenkins (it's also a good exercise!), here is some help to navigate there:
https://github.com/dmwm/WMCore/wiki/Understanding-Jenkins#unit-tests

and also try to reproduce this error with your docker unit tests.

I haven't looked at it myself, so it could be related, or not ;)

**Comment by todor-ivanov** [comment]
@amaltaro I already did these both - examining jenkins in deep and running the failing test with nose in my private instance, and still cannot relate my changes to the failing test in a module few levels aside.

**Comment by amaltaro** [comment]
Yes, it looks like that file is no longer available; or the web server is having issues to serve that file. I made a new GH issue such that we can fix that unit test in the future: https://github.com/dmwm/WMCore/issues/10367

**Comment by todor-ivanov** [comment]
@amaltaro  it is indeed  an error due to a hard coded url [1] in the test and this one  if you try to browse it yourself it fails with the HTTTP erro 500 - internal server error. And if I need to be 100% strict on the fact that my change has no relation to this error here is what is passed to `urllib` directly (url not touched by my change at all) data obtained with pdb [2] 


[1]
https://cms-service-dqm.web.cern.ch/cms-service-dqm/CAF/certification/Collisions12/8TeV/Reprocessing/Cert_190456-195530_8TeV_08Jun2012ReReco_Collisions12_JSON.txt

[2] 
```
> ~/unittest/unittestdeploy/wmagent/1.4.1.pre5-comp10/sw/slc7_amd64_gcc630/external/py2-future/0.18.2/lib/python2.7/site-packages/future/backports/urllib/request.py(157)urlopen()
-> else: capath = None
(Pdb) l
152  	_opener = None
153  	def urlopen(url, data=None, timeout=socket._GLOBAL_DEFAULT_TIMEOUT, **_3to2kwargs):
154  	    if 'cadefault' in _3to2kwargs: cadefault = _3to2kwargs['cadefault']; del _3to2kwargs['cadefault']
155  	    else: cadefault = False
156  	    if 'capath' in _3to2kwargs: capath = _3to2kwargs['capath']; del _3to2kwargs['capath']
157    	    else: capath = None
158  	    if 'cafile' in _3to2kwargs: cafile = _3to2kwargs['cafile']; del _3to2kwargs['cafile']
159  	    else: cafile = None
160  	    global _opener
161  	    if cafile or capath or cadefault:
162  	        if not _have_ssl:
(Pdb) 
163  	            raise ValueError('SSL support not available')
164  	        context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
165  	        context.options |= ssl.OP_NO_SSLv2
166  	        context.verify_mode = ssl.CERT_REQUIRED
167  	        if cafile or capath:
168  	            context.load_verify_locations(cafile, capath)
169  	        else:
170  	            context.set_default_verify_paths()
171  	        https_handler = HTTPSHandler(context=context, check_hostname=True)
172  	        opener = build_opener(https_handler)
173  	    elif _opener is None:
174  	        _opener = opener = build_opener()
175  	    else:
176  	        opener = _opener
177  ->	    return opener.open(url, data, timeout)


(Pdb) url
'https://cms-service-dqm.web.cern.ch/cms-service-dqm/CAF/certification/Collisions12/8TeV/Reprocessing/Cert_190456-195530_8TeV_08Jun2012ReReco_Collisions12_JSON.txt'
(Pdb) 
```

**Comment by todor-ivanov** [comment]
Sorry @amaltaro - I missed your previous comment. Thanks for creating the issue!

File: src/python/Utils/PortForward.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mapellidario) [reply src/python/Utils/PortForward.py:8]: Yeah, you are definitely correct. And it is also true that in this case there is problem in using the `bytes` provided by py2 by default. 

But we are going to switch to py3 sooner or later, and in py3 bytes are not converted to unicode automatically. Importing `bytes` from python-future with `from builtins import bytes` makes us use a [backported equivalent](https://github.com/PythonCharmers/python-future/blob/master/src/future/types/newbytes.py) of py3 bytes in py2 (which fail for example when it is concatenated to unicode) and this helps detecting beforehand when we implicitly rely on the automatic conversion from bytes to unicode.

So, I am sorry if I have been a bit harsh, I did not mean that your code is broken, I am just trying to remind us that it is possible to spot in py2 some problems that we would face in py3 if we simply use the alternatives from python-future. I agree that it is not necessary in this particular case, but I would still do it just to build the habit and not use the py2 `bytes` , which can give the false impression of full py3-compatibility. Does this make sense to you?

**Comment by amaltaro** [comment]
This is looking much better now, timing is very close between the different approaches as well:
```
(PY2) alannewmacpro:ProductionTools amaltaro$ python timePortForward.py 1000000
Going to make 1000000 url conversions
testDecorator received payload: 1000000
Finished single function call in 1.86168313026 secs

testFuncDecorator received payload: 1000000
Finished single function call in 2.5859568119 secs

testSimpleFunc received payload: 1000000
Finished single function call in 1.98184108734 secs
```

File: src/python/Utils/PortForward.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (todor-ivanov) [reply src/python/Utils/PortForward.py]: Done now! Thanks  @mapellidario

[the remark this answers is not in the document]
Comment (mapellidario) [reply src/python/Utils/PortForward.py]: Perfect! thank you, Todor!

**Review by mapellidario — APPROVED** [review]
Thanks Todor for the changes! :) Looks good to me now!

**Review by amaltaro — APPROVED** [review]
Thanks Todor, it looks good to me.

**Pull request merged**